### PR TITLE
Change rpn classification loss layer.

### PR DESCRIPTION
**Pull request by hgaiser**
As mentioned [here](https://github.com/broadinstitute/keras-rcnn/pull/41#discussion_r131342299), this is what I propose. I have not tested this, so I don't know if it is working or not. This PR only changes the classification loss, I didn't touch the regression loss.

Basically what this does is:

1. Reshape predicted classifications to `(None, 2)` (`rpn_labels` is already `(None,)`.
2. Select those classifications that have a label which is not equal to `-1`. These are labelled as background (`0`) or foreground (`1`). `-1` is reserved for "don't care".
3. Similarly, filter out `rpn_labels` with `-1`.
4. Execute `categorical_crossentropy` on the resulting two tensors (`rpn_labels.shape = (None,)` and `rpn_classification.shape = (None, 2)`).

I would love to hear some feedback on this approach of RPN losses.

### CODE REVIEW

File: keras_rcnn/models.py [hunk 1]
```diff
@@ -33,12 +33,15 @@ def __init__(self, inputs, features, heads, rois):
         proposals                        = keras_rcnn.layers.ObjectProposal(maximum_proposals=rois)([im_info, rpn_regression, rpn_classification])
         rpn_labels, rpn_bbox_reg_targets = keras_rcnn.layers.ProposalTarget()([gt_boxes, im_info])
 
+        rpn_classification_loss = keras_rcnn.layers.ClassificationLoss(anchors=num_anchors, name="rpn_classification_loss")([rpn_labels, rpn_classification])
+        #rpn_regression_loss     = keras_rcnn.layers.RegressionLoss(name="rpn_bbox_loss")([rpn_bbox_reg_targets, rpn_regression])
```
Comment (0x00b1) [inline keras_rcnn/models.py]: Did you intend to comment this out? It looks like the correct interface.

File: keras_rcnn/layers/losses.py [hunk 2]
```diff
@@ -1,45 +1,38 @@
...
+        super(ClassificationLoss, self).__init__(**kwargs)
 
-    def _loss(self, y_true, y_pred):
-        # Binary classification loss
-        x, y = y_pred[:, :, :, :], y_true[:, :, :, self.anchors:]
+    def _loss(self, rpn_labels, rpn_classification):
+        rpn_classification = keras.backend.reshape(rpn_classification, [-1, 2])
+        rpn_classification = keras.backend.gather(rpn_classification, keras_rcnn.backend.where(keras.backend.not_equal(rpn_labels, -1)))
+        rpn_labels         = keras.backend.gather(rpn_labels, keras_rcnn.backend.where(keras.backend.not_equal(rpn_labels, -1)))
+        loss               = keras.backend.mean(keras.backend.categorical_crossentropy(rpn_labels, rpn_classification))
```
Comment (0x00b1) [inline keras_rcnn/layers/losses.py]: LGTM

File: keras_rcnn/layers/losses.py [hunk 3]
```diff
@@ -1,45 +1,38 @@
...
-        super(Classification, self).__init__(**kwargs)
+        super(ClassificationLoss, self).__init__(**kwargs)
 
-    def _loss(self, y_true, y_pred):
-        # Binary classification loss
-        x, y = y_pred[:, :, :, :], y_true[:, :, :, self.anchors:]
+    def _loss(self, rpn_labels, rpn_classification):
+        rpn_classification = keras.backend.reshape(rpn_classification, [-1, 2])
+        rpn_classification = keras.backend.gather(rpn_classification, keras_rcnn.backend.where(keras.backend.not_equal(rpn_labels, -1)))
+        rpn_labels         = keras.backend.gather(rpn_labels, keras_rcnn.backend.where(keras.backend.not_equal(rpn_labels, -1)))
```
Comment (0x00b1) [inline keras_rcnn/layers/losses.py]: From @yhenon (on Slack), author of the original loss:

> @hgaiser it looks nice but you should explicitly handle the case where `keras.backend.not_equal(rpn_labels, -1)` returns an empty array (it can happen, especially if rpn_max_proposals is low) - otherwise you will get NaNs from the `mean()` calculation (edited)

File: keras_rcnn/models.py (hunk 1 above)
Comment (hgaiser) [inline keras_rcnn/models.py]: I did, yeah. It was giving me errors because of wrong shapes etc. Figured lets discuss it first, then apply the same logic to all losses :)

File: keras_rcnn/layers/losses.py (hunk 3 above)
Comment (hgaiser) [inline keras_rcnn/layers/losses.py]: I will look into this, makes sense :)

File: keras_rcnn/layers/losses.py (hunk 2 above)
Comment (yhenon) [inline keras_rcnn/layers/losses.py]: This should be `sparse_categorical_crossentropy` not `categorical_crossentropy`. Since your inputs are of shape rpn_labels.shape = (None,) and rpn_classification.shape = (None, 2).

Comment (jhung0) [inline keras_rcnn/layers/losses.py]: I'll merge it as is first and then make this change
